### PR TITLE
WIP net/tcp: make header option pointer c++ compliant

### PR DESCRIPTION
**Pull request by llueder**
### Contribution description

Same as #11299 
I would like to keep discussions which are not specific for this tcp case over at #11299

### CODE REVIEW

**Comment by llueder** [comment]
Ping @brummer-simon  since you are the author of the affected file

File: sys/include/net/tcp.h [hunk 1]
```diff
@@ -72,9 +72,19 @@ typedef struct __attribute__((packed)) {
...
 } tcp_hdr_opt_t;
 
+/**
+ * @brief   Get the start of the TCP option value
+ *
+ * @param[in]   opt   TCP option field helper structure contiguous memory
+ *
+ * @returns     pointer to first byte of the value
+ */
+inline static uint8_t *tcp_hdr_opt_value_ptr(tcp_hdr_opt_t *opt){
```
Comment (brummer-simon) [inline sys/include/net/tcp.h]: Why is this static? I don't see any point in having a static function outside of a compilation unit.

File: sys/net/gnrc/transport_layer/tcp/gnrc_tcp_option.c [hunk 2]
```diff
@@ -55,7 +55,7 @@ int _option_parse(gnrc_tcp_tcb_t *tcb, tcp_hdr_t *hdr)
                     DEBUG("gnrc_tcp_option.c : _option_parse() : invalid MSS Option length.\n");
                     return -1;
                 }
-                tcb->mss = (option->value[0] << 8) | option->value[1];
+                tcb->mss = (tcp_hdr_opt_value_ptr(option)[0] << 8) | tcp_hdr_opt_value_ptr(option)[1];
```
Comment (brummer-simon) [inline sys/net/gnrc/transport_layer/tcp/gnrc_tcp_option.c]: I don't like find the style of "func()[0]" very readable and the calculation is redundant. Please store the Pointer in a variable and use it. I think I would be much clearer to read.

File: sys/include/net/tcp.h (hunk 1 above)
Comment (miri64) [reply sys/include/net/tcp.h]: Making inline functions non-static will cause the linker to fail once you include `net/tcp.h` in multiple c-files ;-).

Comment (miri64) [inline sys/include/net/tcp.h]: Please use the order `static inline`.

**Comment by llueder** [comment]
> I am fine with replacing the dynamic sized array in general, but I wonder since then this is not iso-c++ compliant?

Using cppreference.com as something that should com close to the standard, [classes](https://en.cppreference.com/w/cpp/language/classes) can have [non-static data members](https://en.cppreference.com/w/cpp/language/data_members), which are not allowed to be of [incomplete type](https://en.cppreference.com/w/cpp/language/type#Incomplete_type), for which an [array of unknown bound](https://en.cppreference.com/w/cpp/language/array#Arrays_of_unknown_bound) is an example.
That does technically not answer your question, but it seems like this has always been the case.
I note that gcc can handle it, but issues a warning, what leads to an error in the default build settings of RIOT

File: sys/include/net/tcp.h (hunk 1 above)
Comment (brummer-simon) [reply sys/include/net/tcp.h]: Okay. I just don't understand why. Inlining should be enough to deal with linkage issues.

**Comment by brummer-simon** [comment]
> > I am fine with replacing the dynamic sized array in general, but I wonder since then this is not iso-c++ compliant?
> 
> Using cppreference.com as something that should com close to the standard, [classes](https://en.cppreference.com/w/cpp/language/classes) can have [non-static data members](https://en.cppreference.com/w/cpp/language/data_members), which are not allowed to be of [incomplete type](https://en.cppreference.com/w/cpp/language/type#Incomplete_type), for which an [array of unknown bound](https://en.cppreference.com/w/cpp/language/array#Arrays_of_unknown_bound) is an example.
> That does technically not answer your question, but it seems like this has always been the case.
> I note that gcc can handle it, but issues a warning, what leads to an error in the default build settings of RIOT

Thanks for the links. I learned something. I am just wondering I am using gcc at work and I never notices an warning from such constructs.

**Comment by llueder** [comment]
I pushed the approach that was agreed upon in #11299.

**Comment by brummer-simon** [comment]
Have you tested gnrc_tcp the compiled with c++?

**Comment by llueder** [comment]
Yes that works. I compiled the gnrc_tcp_client test with c++ and ran it against gnrc_tcp_server.
See [this patch](https://gist.github.com/llueder/72604b607de748a660c627c19dea775e), which renames main.c to main.cpp, adds a class to ensure that we're using c++ and applies another fix for c++ use.
I guess the compilation units with the internals are compiled in C.

**Pull request closed without merging**